### ISSUE TITLE
FP `deref_addrof`

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#12319]
### Summary

.

### Lint Name

deref_addrof

### Reproducer

I tried this code:

```rust
#![warn(clippy::deref_addrof)]

fn no_lint_ops() {
    let x = *&mut false |= true && return;
}

pub fn main() {}

```

I saw this happen:

```
warning: immediately dereferencing a reference
 --> src/main.rs:4:13
  |
4 |     let x = *&mut false |= true && return;
  |             ^^^^^^^^^^^ help: try: `false`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#deref_addrof
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(clippy::deref_addrof)]
  |         ^^^^^^^^^^^^^^^^^^^^
```

I expected to see this happen:
it no compily 

```

The following errors were reported:
error[E0067]: invalid left-hand side of assignment
 --> src/main.rs:4:11
  |
4 |     false |= true && return;
  |     ----- ^^
  |     |
  |     cannot assign to this expression

error: aborting due to 1 previous error
```

### Version

```text
rustc 1.78.0-nightly (3246e7951 2024-02-19)
binary: rustc
commit-hash: 3246e79513cb89ddbfc0f21cb5a877e5b321dcc5
commit-date: 2024-02-19
host: x86_64-unknown-linux-gnu
release: 1.78.0-nightly
LLVM version: 18.1.0
```

**Comment by alexis-langlet** [comment]
From what I've seen, `*&mut` is doing a stack allocation and so the `*&mut false` is considered a variable

I was able to confirm the same issue arise using a variable
```rust
fn main() {
    let _ = false && return; // raise error
    let x = false; 
    let _ = x && return; // do not raise error
    let _ = *&mut false && return; // do not raise error
}
```

So I suppose that those checks are not done on variables and that the `*&mut` is "faking" a variable

**Comment by blyxyas** [comment]
I can confirm Alexis' theory. Here's the HIR for that `*&mut false`

```
Expr {
                            hir_id: HirId(DefId(0:3 ~ playground[be21]::no_lint_ops).3),
                            kind: Unary(
                                Deref,
                                Expr {
                                    hir_id: HirId(DefId(0:3 ~ playground[be21]::no_lint_ops).4),
                                    kind: AddrOf(
                                        Ref,
                                        Mut,
                                        Expr {
                                            hir_id: HirId(DefId(0:3 ~ playground[be21]::no_lint_ops).5),
                                            kind: Lit(
                                                Spanned {
                                                    node: Bool(
                                                        false,
                                                    ),
                                                    span: src/main.rs:5:19: 5:24 (#0),
                                                },
                                            ),
                                            span: src/main.rs:5:19: 5:24 (#0),
                                        },
                                    ),
                                    span: src/main.rs:5:14: 5:24 (#0),
                                },
                            ),
                            span: src/main.rs:5:13: 5:24 (#0),
                        },
```

I'll do some bugfixing.
@rustbot claim

**Comment by bluss** [comment]
Also relevant, previous discussion of ndarray applications of *&mut: https://internals.rust-lang.org/t/assign-ops-papercut-with-custom-view-types/
